src/npm/autoupdate/autoupdate.ts

```diff
@@ -3,98 +3,25 @@
-import { ScanPaginator } from "@aws/dynamodb-query-iterator";
-import * as DynamoDB from "aws-sdk/clients/dynamodb";
-import { getRegistryEntryForPackage } from "../util/registry";
+import { S3Cache } from "../S3Cache";
+import { NpmPackage } from "../NpmPackage";
 import "source-map-support/register";
-import * as util from "util";
-import * as zlib from "zlib";
 
-const DDBDocClient = new DynamoDB.DocumentClient();
-const zlibInflatePromise = util.promisify(zlib.inflate);
-const tableName = process.env.DDB_TABLE!;
 const NPM_CACHE_DOWNLOAD_URI = process.env.NPM_CACHE_DOWNLOAD_URI!;
 
-/** Parse compressed DDB entry */
-async function parseDdbEntry(compressedData: any): Promise<any> {
-  const inflatedData: any = await zlibInflatePromise(compressedData);
-  return JSON.parse(inflatedData.toString());
-}
-
-/** Update DDB entry with new packages */
-async function updateDdbEntry(entry: any): Promise<any> {
-  var packageName = entry.PackageName;
-  var downloadUriPrefix = NPM_CACHE_DOWNLOAD_URI + "/" + packageName + "/";
-
-  // Retrieve upstream registryData & DDB cached ddbData
-  var [registryData, ddbData] = (await Promise.all([
-    getRegistryEntryForPackage(packageName),
-    parseDdbEntry(entry["CompressedRegistryData"]),
-  ])) as any;
-
-  if (Date.parse(ddbData["modified"]) < Date.parse(registryData["modified"])) {
-    console.log(
-      "Package " +
-        packageName +
-        " has been modified upstream; updating its ddb entry",
-    );
-    Object.keys(registryData).forEach(function(value, _index, _array) {
-      if (value === "versions") {
-        Object.keys(registryData["versions"]).forEach(function(
-          versionNumber,
-          _versionIndex,
-          _versionArray,
-        ) {
-          // Leaving existing entries untouched.
-          if (!(versionNumber in ddbData["versions"])) {
-            console.log(
-              "Adding new version " +
-                versionNumber +
-                " to " +
-                packageName +
-                " entry.",
-            );
-            ddbData["versions"][versionNumber] =
-              registryData["versions"][versionNumber];
-            // update download links to point to proxy endpoint
-            ddbData["versions"][versionNumber]["dist"]["tarball"] =
-              downloadUriPrefix + versionNumber;
-          }
-        });
-      } else {
-        ddbData[value] = registryData[value];
-      }
-    });
-    await DDBDocClient.put({
-      TableName: tableName,
-      Item: {
-        PackageName: packageName,
-        CompressedRegistryData: zlib.deflateSync(JSON.stringify(ddbData)),
-      },
-    }).promise();
-  }
-}
-
 /** AWS Lambda entrypoint */
 export let handler: Handler = async (
   event: ScheduledEvent,
   context: Context,
 ): Promise<any> => {
-  const paginator = new ScanPaginator((DDBDocClient as unknown) as DynamoDB, {
-    TableName: tableName,
-  });
-
+  const s3Cache = new S3Cache();
```

Comment:
We should aim to instantiate as many objects as possible outside of the handler, so they get reused on subsequent invocations.

The most notable ones being the AWS client objects, like the S3 client in S3Cache here.

Reply:
are you OK if we add an IOC (inversion of control) lib?. e.g. http://inversify.io/
https://en.wikipedia.org/wiki/Inversion_of_control

Reply:
My preference would be against it (the focus in this project is simplicity & speed), but I'm not wholly opposed to it. Would need to be [sailplane injector](https://github.com/onicagroup/sailplane/tree/master/injector) (or bottlejs directly) to match our organizational preference/experience.

Reply:
Agreed, I'll create all instances in the lambda handler. Thanks!

Reply:
fixed in https://github.com/troyready/serverless-artifact-cacher/pull/3/commits/bd79d9d2e98691c0219e903884666ad37f570966